python/ray/dag/dag_node_operation.py

```diff
@@ -0,0 +1,337 @@
+    to generate the schedule.
+
+    Args:
+        graph: A graph where each node is a _DAGOperationGraphNode. The key is
+            `dag_idx`, the index to retrieve its task from `idx_to_task`, and
+            the value is a dictionary that maps the _DAGNodeOperationType (READ,
+            COMPUTE, or WRITE) to the corresponding _DAGOperationGraphNode. It is
+            generated by `_build_dag_node_operation_graph`.
+
+    Returns:
+        actor_to_execution_schedule: A dictionary that maps an actor handle to
+            the execution schedule which is a list of operations to be executed.
+    """
+
+    # Mapping from the actor handle to the execution schedule which is a list
+    # of operations to be executed.
+    actor_to_execution_schedule: Dict[
+        "ray.actor.ActorHandle", List[_DAGNodeOperation]
+    ] = defaultdict(list)
+
+    # A dictionary mapping an actor id to a list of candidate nodes. The list
+    # is maintained as a priority queue, so the head of the queue, i.e.,
+    # `candidates[0]`, is the node with the smallest `bind_index`.
+    actor_to_candidates: Dict[
+        "ray._raylet.ActorID", List[_DAGOperationGraphNode]
+    ] = defaultdict(list)
+    for _, node_dict in graph.items():
+        for _, node in node_dict.items():
+            # A node with a zero in-degree edge means all of its dependencies
+            # have been satisfied, including both data and control dependencies.
+            # Therefore, it is a candidate for execution.
+            if node.in_degree == 0:
+                heapq.heappush(actor_to_candidates[node.actor_handle._actor_id], node)
+
+    visited_nodes = set()
+
+    # Topological sort
+    while actor_to_candidates:
+        # The function `_select_next_nodes` will pop a candidate node from
+        # `actor_to_candidates` and return a list of nodes that can be executed
+        # in the next step. If multiple nodes are returned, only the NCCL write
+        # node is popped in this iteration.
+        nodes = _select_next_nodes(actor_to_candidates, graph)
+        for node in nodes:
+            if node in visited_nodes:
+                continue
+            actor_to_execution_schedule[node.actor_handle].append(node.operation)
+            visited_nodes.add(node)
+            for out_node_dag_idx, out_node_type in node.out_edges:
+                out_node = graph[out_node_dag_idx][out_node_type]
+                out_node.in_edges.remove((node.dag_idx, node.operation.type))
+                if out_node.in_degree == 0:
+                    heapq.heappush(
+                        actor_to_candidates[out_node.actor_handle._actor_id],
+                        out_node,
+                    )
+
+        delete_keys = []
+        for actor_id, candidates in actor_to_candidates.items():
+            if len(candidates) == 0:
+                delete_keys.append(actor_id)
+        for key in delete_keys:
+            del actor_to_candidates[key]
```

Comment:
Please remove this code; it is not necessary and makes the code more complicated than necessary (you can just check `if actor_to_candidates` in `_select_next_nodes`).

Reply:
Updated https://github.com/ray-project/ray/pull/46911/commits/d97fda6fd508d4dbc7da893d25ca4b1dd830e20d